### PR TITLE
Users/shigupt/kubernetes manifest bug fixes

### PR DESCRIPTION
**Pull request by shigupt202**
1. canaryPercentage and baselineAndCanaryReplicas validation added
2. Added check for baseline and canary resources list before adding baseline and canary resources to deletion list in promote/reject
3. Made 'namespace' an optional field
4. Removed hard-coding for the api version used in TrafficSplit resource.

### CODE REVIEW

File: Tasks/KubernetesManifestV0/src/actions/deploy.ts [hunk 1]
```diff
@@ -6,6 +6,8 @@ import * as utils from '../utils/utilities';
 import { Kubectl } from 'kubernetes-common-v2/kubectl-object-model';
 
 export async function deploy(ignoreSslErrors?: boolean) {
+    TaskInputParameters.validateCanaryPercentage();
+    TaskInputParameters.validateReplicaCount();
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/actions/deploy.ts]: You could skip this getting called altogether. By calling the logic in the file itself.

File: Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts [hunk 2]
```diff
@@ -26,4 +27,25 @@ if (!namespace) {
...
     tl.debug('Namespace was not supplied nor present in the endpoint; using "default" namespace instead.');
     namespace = 'default';
-}
\ No newline at end of file
+}
+
+export function validateCanaryPercentage()
+{
+    if(deploymentStrategy.toUpperCase() === canaryDeploymentHelper.CANARY_DEPLOYMENT_STRATEGY && !validateRegex("^(([0-9]|[1-9][0-9]|100)(\\.\\d{1,2})?)$", canaryPercentage))
+    {
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts]: You could use ParseInt function right?

File: Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts [hunk 3]
```diff
@@ -26,4 +27,25 @@ if (!namespace) {
...
+    if(deploymentStrategy.toUpperCase() === canaryDeploymentHelper.CANARY_DEPLOYMENT_STRATEGY && trafficSplitMethod.toUpperCase() === canaryDeploymentHelper.TRAFFIC_SPLIT_STRATEGY && !validateRegex("^(([0-9]|[1-9][0-9]|100)(\\.\\d{1,2})?)$", baselineAndCanaryReplicas))
+    {
+        throw new Error(tl.loc('InvalidBaselineAndCanaryReplicas'));
+    }
+}
+
+function validateRegex(regex: string, testString: string) {
+    var percentageRegex = new RegExp(regex);
+    return percentageRegex.test(testString);
+}
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts]: Run the vscode auto formatter to fix the indents for consistency

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts [hunk 4]
```diff
@@ -32,17 +33,7 @@ export function deleteCanaryDeployment(kubectl: Kubectl, manifestFilePaths: stri
...
-    // append delete cmd args as suffix (if present)
-    const args = utils.getDeleteCmdArgs(argsPrefix, TaskInputParameters.args);
-    tl.debug('Delete cmd args : ' + args);
-
-    if (!!args && args.length > 0) {
-        // run kubectl delete cmd
-        const result = kubectl.delete(args);
-        utils.checkForErrors([result]);
-    }
+    createCanaryObjectsArgumentString(kubectl, inputManifestFiles, includeServices);
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: Rename the function and remove the declaration of the above variable

File: Tasks/KubernetesManifestV0/src/utils/utilities.ts [hunk 5]
```diff
@@ -147,7 +148,21 @@ export function substituteImageNameInSpecFile(currentString: string, imageName:
     }, '');
 }
 
+export function getAPIVersion() {
+    if(apiVersion)
+        return apiVersion;
+
+    const kubectlPath = tl.which('kubectl', true)
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/utilities.ts]: A Kubectl path should be floating around in the action, reuse that.

File: Tasks/KubernetesManifestV0/src/utils/utilities.ts [hunk 6]
```diff
@@ -147,7 +148,21 @@ export function substituteImageNameInSpecFile(currentString: string, imageName:
...
+    const command = tl.tool(kubectlPath);
+    command.arg('api-versions');
+    const result = command.execSync().stdout.split('\n');
+    apiVersion = result.find(version => version.startsWith('split.smi-spec.io'));
+
+    return apiVersion = (apiVersion != null && typeof apiVersion != 'undefined') ? apiVersion : 'split.smi-spec.io/v1alpha1';
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/utilities.ts]: If it was unable wouldn't the deployment fail?

Comment (ds-ms) [reply Tasks/KubernetesManifestV0/src/utils/utilities.ts]: Throw a warning otherwise

File: Tasks/KubernetesManifestV0/task.json [hunk 7]
```diff
@@ -340,30 +340,32 @@
...
+    "CouldNotDeterminePodStatus": "Could not determine the pod's status due to the error: %s",
+    "KubectlShouldBeUpgraded": "kubectl client version equal to v1.14 or higher is required to use kustomize features.",
+    "CouldNotDetermineServiceStatus": "Could not determine the service %s status due to the error: %s",
+    "waitForServiceIpAssignment": "Waiting for service %s external IP assignment",
+    "waitForServiceIpAssignmentTimedOut": "Wait for service %s external IP assignment timed out",
+    "ServiceExternalIP": "service %s external IP is %s",
+    "UnableToCreateTrafficSplitManifestFile": "Unable to create TrafficSplit manifest file.",
+    "StableSpecSelectorNotExist": "Resource %s not deployed using SMI canary deployment.",
+    "InvalidPercentage": "Invalid value for percentage.",
+    "InvalidBaselineAndCanaryReplicas":  "Invalid value for replica count. Enter a value between 0 and 100."
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/task.json]: Why's the indent different?

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts [hunk 8]
```diff
@@ -207,17 +201,29 @@ function createCanaryObjectsArgumentString(files: string[], includeServices: boo
                 || (includeServices && helper.isServiceEntity(kind))) {
                 const canaryObjectName = getCanaryResourceName(name);
                 const baselineObjectName = getBaselineResourceName(name);
-                kindList.add(kind);
-                nameList.add(canaryObjectName);
-                nameList.add(baselineObjectName);
+                addValueToList(kindNameMap, kind, canaryObjectName);
+                const result = kubectl.getResource(kind, baselineObjectName);
+                if(result != null && !result.stderr)
+                {
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: Have the same formatting style across the files

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts [hunk 9]
```diff
@@ -207,17 +201,29 @@ function createCanaryObjectsArgumentString(files: string[], includeServices: boo
...
+                    addValueToList(kindNameMap, kind, baselineObjectName);
+                }
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: Why don't you run the Kubectl delete logic here itself? You don't need to maintain a map then

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts [hunk 10]
```diff
@@ -10,6 +10,7 @@ import * as helper from '../utils/KubernetesObjectUtility';
 import { KubernetesWorkload } from 'kubernetes-common-v2/kubernetesconstants';
 import { StringComparer, isEqual } from './StringComparison';
 import * as utils from '../utils/utilities';
+import { isFulfilled } from 'q';
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: This import is not needed right?

File: Tasks/KubernetesManifestV0/src/actions/deploy.ts (hunk 1 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/actions/deploy.ts]: If the call is made in the same file, the errors thrown by these validation functions will not be caught by the catch block in run.ts and will remain unhandled.

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts (hunk 4 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: Okay

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts (hunk 9 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: This is done to reduce the number of API calls.

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts (hunk 10 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: This is not needed.

File: Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts (hunk 2 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts]: Percentage should be a floating value.

File: Tasks/KubernetesManifestV0/src/utils/utilities.ts (hunk 5 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/utilities.ts]: Kubectl path has been used only once per action to extract the kubectl object. If I try to reuse that, a lot of function calls will have to be changed.

File: Tasks/KubernetesManifestV0/src/utils/utilities.ts (hunk 6 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/utilities.ts]: I'll add that

File: Tasks/Common/kubernetes-common-v2/kubernetesmanifestutility.ts [hunk 11]
```diff
@@ -9,10 +9,10 @@ import { Kubectl, Resource } from './kubectl-object-model';
 export async function checkManifestStability(kubectl: Kubectl, resources: Resource[]): Promise<void> {
     const rolloutStatusResults = [];
     const numberOfResources = resources.length;
-    for (let i = 0; i< numberOfResources; i++) {
+    for (let i = 0; i < numberOfResources; i++) {
         const resource = resources[i];
         if (KubernetesConstants.workloadTypesWithRolloutStatus.indexOf(resource.type.toLowerCase()) >= 0) {
-            rolloutStatusResults.push(kubectl.checkRolloutStatus(resource.type, resource.name));
+            rolloutStatusResults.push(kubectl.checkRolloutStatus(resource.type, resource.name, tl.getInput('timeout')));
```
Comment (ds-ms) [inline Tasks/Common/kubernetes-common-v2/kubernetesmanifestutility.ts]: Make this value come from the caller

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts [hunk 12]
```diff
@@ -207,17 +199,28 @@ function createCanaryObjectsArgumentString(files: string[], includeServices: boo
                 || (includeServices && helper.isServiceEntity(kind))) {
                 const canaryObjectName = getCanaryResourceName(name);
                 const baselineObjectName = getBaselineResourceName(name);
-                kindList.add(kind);
-                nameList.add(canaryObjectName);
-                nameList.add(baselineObjectName);
+                addValueToList(kindNameMap, kind, canaryObjectName);
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: The assumption is that `canary` will always exist?

Comment (ds-ms) [reply Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: People could accidentally use canary first too; do you want to handle that case?

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts [hunk 13]
```diff
@@ -194,10 +183,13 @@ function addCanaryLabelsAndAnnotations(inputObject: any, type: string) {
     }
 }
 
-function createCanaryObjectsArgumentString(files: string[], includeServices: boolean) {
-    const kindList = new Set();
-    const nameList = new Set();
+function addValueToList(map, key, value) {
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: Given that the intent is to optimize the calls; you could convert the key to lowercase. 
So, when there are manifests with kind `Deployment` and `deployment` the calls would be optimized to a single one instead of two

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts (hunk 12 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: This function is called only for promote and reject. So, we can assume that canary deployment will always be present.

File: Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts (hunk 13 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/src/utils/CanaryDeploymentHelper.ts]: Okay

File: Tasks/KubernetesManifestV0/task.json [hunk 14]
```diff
@@ -325,16 +324,22 @@
             "label": "Docker registry service connection",
             "helpMarkDown": "Select a Docker registry service connection. Required for commands that need to authenticate with a registry.",
             "visibleRule": "action = createSecret && secretType = dockerRegistry"
+        },
+        {
+            "name": "timeout",
+            "tpye": "string",
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/task.json]: Should be type

File: Tasks/KubernetesManifestV0/task.json [hunk 15]
```diff
@@ -325,16 +324,22 @@
...
+            "label": "Timeout for rollout status",
+            "required": false,
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/task.json]: Add default of 5m

File: Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts [hunk 16]
```diff
@@ -15,6 +16,7 @@ export const secretArguments: string = tl.getInput('secretArguments', false) ||
 export const secretType: string = tl.getInput('secretType', false);
 export const secretName: string = tl.getInput('secretName', false);
 export const dockerRegistryEndpoint: string = tl.getInput('dockerRegistryEndpoint', false);
+export const timeout: string = tl.getInput('timeout', false);
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts]: It's retrieved as a string input. Check if type cast to integer succeeds. If false, you can throw an error here that the value needs be an integer

File: Tasks/KubernetesManifestV0/task.loc.json [hunk 17]
```diff
@@ -325,16 +324,22 @@
       "label": "ms-resource:loc.input.label.dockerRegistryEndpoint",
       "helpMarkDown": "ms-resource:loc.input.help.dockerRegistryEndpoint",
       "visibleRule": "action = createSecret && secretType = dockerRegistry"
+    },
+    {
+      "name": "timeout",
+      "tpye": "string",
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/task.loc.json]: Should be type

File: Tasks/KubernetesManifestV0/Strings/resources.resjson/en-US/resources.resjson [hunk 18]
```diff
@@ -83,6 +85,8 @@
   "loc.messages.waitForServiceIpAssignment": "Waiting for service %s external IP assignment",
   "loc.messages.waitForServiceIpAssignmentTimedOut": "Wait for service %s external IP assignment timed out",
   "loc.messages.ServiceExternalIP": "service %s external IP is %s",
-  "loc.messages.UnableToCreateTrafficSplitManifestFile": "Unable to create TrafficSplit manifest file.",
-  "loc.messages.StableSpecSelectorNotExist": "Resource %s not deployed using SMI canary deployment."
+  "loc.messages.UnableToCreateTrafficSplitManifestFile": "Unable to create TrafficSplit manifest file. %s",
+  "loc.messages.StableSpecSelectorNotExist": "Resource %s not deployed using SMI canary deployment.",
+  "loc.messages.InvalidPercentage": "Invalid value for percentage.",
+  "loc.messages.InvalidBaselineAndCanaryReplicas": "Invalid value for replica count. Enter a value between 0 and 100."
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/Strings/resources.resjson/en-US/resources.resjson]: It just needs to be > 0. No restriction that it should be less than 100

File: Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts [hunk 19]
```diff
@@ -26,4 +28,21 @@ if (!namespace) {
 if (!namespace) {
     tl.debug('Namespace was not supplied nor present in the endpoint; using "default" namespace instead.');
     namespace = 'default';
-}
\ No newline at end of file
+}
+
+export function validateCanaryPercentage() {
+    if (deploymentStrategy.toUpperCase() === canaryDeploymentHelper.CANARY_DEPLOYMENT_STRATEGY && !validateRegex("^(([0-9]|[1-9][0-9]|100)(\\.\\d{1,2})?)$", canaryPercentage)) {
```
Comment (shashankbarsin) [inline Tasks/KubernetesManifestV0/src/models/TaskInputParameters.ts]: Should be > 0% and it's not necessary that decimal places be provided only upto 2 places

File: Tasks/KubernetesManifestV0/task.json [hunk 20]
```diff
@@ -88,7 +87,7 @@
             "defaultValue": 0,
             "visibleRule": "strategy = Canary && action = deploy",
             "validation": {
-                "expression": "isMatch(value, '(^(([0-9]|[1-9][0-9]|100)(\\.\\d{1,2})?)$)','Multiline')",
+                "expression": "isMatch(value, '(^(([0-9]|[1-9][0-9]|100)(\\.\\d*)?)$)','Multiline')",
```
Comment (ds-ms) [inline Tasks/KubernetesManifestV0/task.json]: Why is this change required? is it blocking 100?

File: Tasks/KubernetesManifestV0/task.json (hunk 15 above)
Comment (ds-ms) [reply Tasks/KubernetesManifestV0/task.json]: Default value could be set to 0 where you would take the kubectl's default value

File: Tasks/KubernetesManifestV0/task.json (hunk 20 above)
Comment (shigupt202) [reply Tasks/KubernetesManifestV0/task.json]: It's not blocking 100. But, the previous rule did not allow more than 2 digits after decimal. So values like 95.125 won't be accepted.

File: Tasks/KubernetesManifestV0/task.json (hunk 15 above)
Comment (shashankbarsin) [reply Tasks/KubernetesManifestV0/task.json]: Keep default as whatever value we are using for the stability checks in the task today. Setting any other value will be an unexpected change for whoever is already using the task

Comment (shigupt202) [reply Tasks/KubernetesManifestV0/task.json]: By default, if the deployment is not successful even after 10 min, we get an error stating that the progress deadline exceeded. Setting the default value as 0 means that the watch on rollout status would be indefinitely true, and thus after 10 min it would show the same error.
